Source/wayland/WaylandServer.m

```diff
@@ -0,0 +1,1482 @@
+			       buttonNumber: 0 /* FIXME */
+				     deltaX: deltaX
+				     deltaY: deltaY
+				     deltaZ: 0.];
+
+	[GSCurrentServer() postEvent: event atStart: NO];
+    }
+
+    wlconfig->pointer.x = sx;
+    wlconfig->pointer.y = sy;
+}
+
+static void
+pointer_handle_button(void *data, struct wl_pointer *pointer, uint32_t serial,
+		      uint32_t time, uint32_t button, uint32_t state_w)
+{
+    NSDebugLog(@"pointer_handle_button: button=%d", button);
+    WaylandConfig *wlconfig = data;
+    NSEvent *event;
+    NSEventType eventType;
+    NSPoint eventLocation;
+    NSGraphicsContext *gcontext;
+    unsigned int eventFlags;
+    float deltaX = 0.0;
+    float deltaY = 0.0;
+    int clickCount = 1;
+    int tick;
+    enum wl_pointer_button_state state = state_w;
+    struct window *window = wlconfig->pointer.focus;
+
+    gcontext = GSCurrentContext();
+    eventLocation = NSMakePoint(wlconfig->pointer.x,
+				window->height - wlconfig->pointer.y);
+    eventFlags = 0;
+    if (state == WL_POINTER_BUTTON_STATE_PRESSED) {
+	if (button == wlconfig->pointer.last_click_button &&
+	    time - wlconfig->pointer.last_click_time < 300 &&
+	    abs(wlconfig->pointer.x - wlconfig->pointer.last_click_x) < 3 &&
+	    abs(wlconfig->pointer.y - wlconfig->pointer.last_click_y) < 3) {
+	    NSDebugLog(@"handle_button HIT: b=%d t=%d x=%f y=%f", button, time, wlconfig->pointer.x, wlconfig->pointer.y);
+	    wlconfig->pointer.last_click_time = 0;
+	    clickCount++;
+	} else {
+	    NSDebugLog(@"handle_button MISS: b=%d t=%d x=%f y=%f", button, time, wlconfig->pointer.x, wlconfig->pointer.y);
+	    wlconfig->pointer.last_click_button = button;
+	    wlconfig->pointer.last_click_time = time;
+	    wlconfig->pointer.last_click_x = wlconfig->pointer.x;
+	    wlconfig->pointer.last_click_y = wlconfig->pointer.y;
+	}
+
+	switch (button) {
+	case BTN_LEFT:
+	    eventType = NSLeftMouseDown;
+	    break;
+	case BTN_RIGHT:
+	    eventType = NSRightMouseDown;
+	    break;
+	}
+	wlconfig->pointer.serial = serial;
+    } else {
+	switch (button) {
+	case BTN_LEFT:
+	    eventType = NSLeftMouseUp;
+	    break;
+	case BTN_RIGHT:
+	    eventType = NSRightMouseUp;
+	    break;
+	}
+	wlconfig->pointer.serial = 0;
+    }
+
+    tick = serial;
+
+    NSDebugLog(@"sending pointer event at: %fx%f, window=%d", wlconfig->pointer.x, wlconfig->pointer.y, window->window_id);
+
+    event = [NSEvent mouseEventWithType: eventType
+			       location: eventLocation
+			  modifierFlags: eventFlags
+			      timestamp: (NSTimeInterval) time / 1000.0
+			   windowNumber: (int)window->window_id
+				context: gcontext
+			    eventNumber: tick
+			     clickCount: clickCount
+			       pressure: 1.0
+			   buttonNumber: 0 /* FIXME */
+				 deltaX: deltaX
+				 deltaY: deltaY
+				 deltaZ: 0.];
+
+    [GSCurrentServer() postEvent: event atStart: NO];
+}
+
+static void
+pointer_handle_axis(void *data, struct wl_pointer *pointer,
+		    uint32_t time, uint32_t axis, wl_fixed_t value)
+{
+}
```

Comment:
This should send scroll wheel events.

Reply:
Done, albeit it feels clumsily. Then again, the X11 implementation is not that beautiful either...